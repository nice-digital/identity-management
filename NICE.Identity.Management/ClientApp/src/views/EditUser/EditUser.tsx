import React, { useEffect, useState } from "react";
import { RouteComponentProps, Link, Redirect } from "react-router-dom";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { FormGroup } from "@nice-digital/nds-form-group";
import { Input } from "@nice-digital/nds-input";
import { Radio } from "@nice-digital/nds-radio";
import { PageHeader } from "@nice-digital/nds-page-header";
import { useFetch } from "../../helpers/useFetch";
import { isDataError } from "../../helpers/isDataError";
import { Endpoints } from "../../data/endpoints";
import { UserType } from "../../models/types";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import "@nice-digital/nds-radio/scss/radio.scss";

type TParams = { id: string };
type EditUserProps = Record<string, unknown> & RouteComponentProps<TParams>;
type EditUserForm = Record<string, HTMLInputElement>;
type CustomError = {
	error: Error;
	status: number;
};

export const EditUser = (props: EditUserProps): React.ReactElement => {
	const { id } = props.match.params;

	const [user, setUser] = useState<UserType>(Object);
	const [isLoading, setIsLoading] = useState(false);
	const [isSaveButtonLoading, setIsSaveButtonLoading] = useState(false);
	const [error, setError] = useState<Error | null>(null);
	const [validationErrorList, setValidationErrorList] = useState<any>({});
	const [emailBlockedPattern, setEmailBlockedPattern] =
		useState<{ pattern: string }>(Object);
	const [emailBlockedMessage, setEmailBlockedMessage] = useState<string>("");
	const [redirect, setRedirect] = useState(false);

	const doFetch = useFetch(Endpoints.user(id));

	useEffect(() => {
		let isMounted = true;

		(async () => {
			if (isMounted) {
				setIsLoading(true);
				const data = await doFetch<UserType>();

				if (containsError(data)) {
					const errorObject = data as CustomError;
					setError(errorObject.error);
				} else {
					setUser(data as UserType);
				}

				setIsLoading(false);
			}
		})();

		return () => {
			isMounted = true;
		};
	}, []);

	const containsError = (data: Record<string, unknown>) => {
		return Object.prototype.hasOwnProperty.call(data, "error");
	};

	const formHasChanged = (form: EditUserForm): boolean => {
		const formValuesAreDifferent = [
			user.emailAddress !== form.emailAddress.value,
			user.firstName !== form.firstName.value,
			user.lastName !== form.lastName.value,
			user.allowContactMe !== form.audienceInsight_optIn.checked,
		];

		if (formValuesAreDifferent.includes(true)) {
			return true;
		}

		return false;
	};

	const handleSubmit = async (
		e: React.FormEvent<HTMLFormElement>,
	): Promise<void | boolean> => {
		e.preventDefault();
		setIsSaveButtonLoading(true);

		const form = e.currentTarget;
		const { emailAddress, firstName, lastName, audienceInsight_optIn } = form;
		const validationErrors = { ...validationErrorList };

		if (!form.checkValidity()) {
			setIsSaveButtonLoading(false);
			return false;
		}

		if (!formHasChanged(form)) {
			setRedirect(true);
			return false;
		}

		const fetchOptions = {
			method: "PATCH",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({
				userId: Number(id),
				emailAddress: emailAddress.value,
				firstName: firstName.value,
				lastName: lastName.value,
				allowContactMe: audienceInsight_optIn.checked,
			}),
		};

		const data = await doFetch<UserType>(Endpoints.user(id), fetchOptions);

		if (!containsError(data)) {
			setUser(data as UserType);
			setRedirect(true);
		} else {
			const errorObject = data as CustomError;

			if (errorObject.status === 422) {
				validationErrors["emailAddressBlocked"] = true;
				setEmailBlockedPattern({
					pattern: updateEmailBlockedPattern(emailAddress.value),
				});
				setEmailBlockedMessage(errorObject.error.message);
				setValidationErrorList(validationErrors);
			} else {
				setError(error);
			}
		}

		setIsSaveButtonLoading(false);
	};

	const updateEmailBlockedPattern = (emailAddress: string): string => {
		const regexPattern = Object.prototype.hasOwnProperty.call(
			emailBlockedPattern,
			"pattern",
		)
			? emailBlockedPattern.pattern.replace("?!(", `?!(${emailAddress}|`)
			: `^(?!(${emailAddress})$).*`;

		return regexPattern;
	};

	const handleChange = (e: React.ChangeEvent<HTMLInputElement>): void => {
		const formElement = e.currentTarget;
		const validationErrors = { ...validationErrorList };

		if (validationErrors[formElement.name] && formElement.validity.valid) {
			validationErrors[formElement.name] = false;
		}

		if (
			formElement.name === "emailAddress" &&
			validationErrors["emailAddressBlocked"]
		) {
			validationErrors["emailAddressBlocked"] = false;
		}

		setValidationErrorList(validationErrors);
	};

	const handleBlur = (e: React.FocusEvent<HTMLInputElement>): void => {
		const formElement = e.currentTarget;
		const validationErrors = { ...validationErrorList };

		validationErrors[formElement.name] = !formElement.validity.valid;

		if (
			formElement.name === "emailAddress" &&
			Object.prototype.hasOwnProperty.call(emailBlockedPattern, "pattern")
		) {
			const pattern = new RegExp(emailDuplicatePattern.pattern);

			validationErrors["emailAddressDuplicate"] = !pattern.test(
				formElement.value,
			)
				? true
				: false;
		}

		setValidationErrorList(validationErrors);
	};

	const usernameBreadcrumb = isLoading
		? "Loading user details"
		: error
		? "Error"
		: `${user.firstName} ${user.lastName}`;

	return (
		<>
			<Breadcrumbs>
				<Breadcrumb
					data-qa-sel="breadcrumb-user-link"
					to="/users"
					elementType={Link}
				>
					Users
				</Breadcrumb>
				<Breadcrumb to={`/users/${id}`} elementType={Link}>
					{usernameBreadcrumb}
				</Breadcrumb>
				<Breadcrumb>Edit user</Breadcrumb>
			</Breadcrumbs>

			{!error ? (
				<>
					{isLoading ? (
						<>
							<PageHeader preheading="Confirm" heading="Edit user" />
							<p>Loading...</p>
						</>
					) : (
						<>
							{redirect && (
								<Redirect
									to={{
										pathname: `/users/${id}`,
										state: { hasBeenEdited: true },
									}}
								/>
							)}

							<PageHeader
								data-qa-sel="page-header-edit-user"
								preheading="Personal details for"
								heading={`${user.firstName} ${user.lastName}`}
							/>
							<Grid>
								<GridItem cols={12} md={6} lg={4}>
									<form onSubmit={handleSubmit} noValidate>
										<Input
											data-qa-sel="email-input-edit-user"
											label="Email address"
											name="emailAddress"
											type="email"
											required
											autoComplete="off"
											{...emailBlockedPattern}
											onBlur={handleBlur}
											onChange={handleChange}
											error={
												validationErrorList.emailAddress ||
												validationErrorList.emailAddressBlocked
											}
											// errorMessage={
											// 	validationErrorList.emailAddressBlocked
											// 		? "Email address is already in use"
											// 		: "Email address is in an invalid format"
											// }
											errorMessage={emailBlockedMessage}
											defaultValue={user.emailAddress}
											disabled={isSaveButtonLoading}
										/>
										<Input
											data-qa-sel="firstname-input-edit-user"
											label="First name"
											name="firstName"
											type="text"
											required
											autoComplete="off"
											minLength="2"
											maxLength="100"
											pattern="^((?![<>]).)*$"
											onBlur={handleBlur}
											onChange={handleChange}
											error={validationErrorList.firstName}
											errorMessage="First name should contain letters and should not exceed 100 characters"
											defaultValue={user.firstName}
											disabled={isSaveButtonLoading}
										/>
										<Input
											data-qa-sel="lastname-input-edit-user"
											label="Last name"
											name="lastName"
											type="text"
											required
											autoComplete="off"
											minLength="2"
											maxLength="100"
											pattern="^((?![<>]).)*$"
											onBlur={handleBlur}
											onChange={handleChange}
											error={validationErrorList.lastName}
											errorMessage="Last name should contain letters and should not exceed 100 characters"
											defaultValue={user.lastName}
											disabled={isSaveButtonLoading}
										/>

										<div className="form-group-container form-group-container--no-border">
											<FormGroup
												legend="Audience insight community membership"
												name="audienceInsight"
												disabled={isSaveButtonLoading}
											>
												<Radio
													data-qa-sel="optIn-radio-edit-user"
													value="optIn"
													label="Opt in"
													defaultChecked={user.allowContactMe}
												/>
												<Radio
													data-qa-sel="optOut-radio-edit-user"
													value="optOut"
													label="Opt out"
													defaultChecked={!user.allowContactMe}
												/>
											</FormGroup>
											<Button
												data-qa-sel="save-button-edit-user"
												variant="cta"
												type="submit"
												disabled={isSaveButtonLoading}
											>
												Save profile
											</Button>
											<Button
												to={`/users/${id}`}
												variant="secondary"
												elementType={Link}
												disabled={isSaveButtonLoading}
											>
												Cancel
											</Button>
										</div>
									</form>
								</GridItem>
							</Grid>
						</>
					)}
				</>
			) : (
				<ErrorMessage error={error} />
			)}
		</>
	);
};
