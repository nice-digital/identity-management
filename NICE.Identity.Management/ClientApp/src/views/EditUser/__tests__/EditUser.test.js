import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import { Redirect } from "react-router-dom";
import toJson from "enzyme-to-json";

import { EditUser } from "../EditUser";
import singleUser from "./singleUser.json";
import { nextTick } from "../../../utils/nextTick";

describe("EditUser", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

    const consoleErrorReset = console.error;

	beforeEach(() => {
		fetch.resetMocks();
		console.error = consoleErrorReset;
	});

    it("should show loading message before data has been loaded", () => {
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		const wrapper = shallow(<EditUser match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

    it("should match the snapshot after data has been loaded", async () => {
        fetch.mockResponseOnce(JSON.stringify(singleUser));
        const wrapper = shallow(<EditUser match={match} />);
        await nextTick();
        wrapper.update();
        expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
    });

    it("should disable form submit button when clicked", async () => {
		console.error = jest.fn();
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify({}));
		const wrapper = mount(
			<MemoryRouter>
				<EditUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").simulate("submit");
		expect(wrapper.find("button").props().disabled).toEqual(true);
		expect(wrapper.find("button").text()).toEqual("Loading...");
	});

	it("should display confirmation message once submit is complete", async () => {
		console.error = jest.fn();
		const singleUserEdited = { ...singleUser };
		singleUserEdited.firstName = "John";
		fetch.mockResponseOnce(JSON.stringify(singleUser));
		fetch.mockResponseOnce(JSON.stringify(singleUserEdited));
		const wrapper = mount(
			<MemoryRouter>
				<EditUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("form").prop("onSubmit")({
			preventDefault: () => false, 
			currentTarget: { 
				checkValidity: () => true
			}
			//target: { emailAddress: { value: "johnz.olland@outlook.com" }},
		});
		await nextTick();
		wrapper.update();
		expect(wrapper.find(Redirect).exists()).toBe(true);
	});

});