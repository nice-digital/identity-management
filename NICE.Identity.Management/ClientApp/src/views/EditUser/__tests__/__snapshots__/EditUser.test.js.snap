// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display alert message if editing an EPPI user 1`] = `
<p>
  This user may have access to EPPI R5 - only a professional email address can be associated to this profile. Please verify via the EPPI user admin page before changing the email address.
</p>
`;

exports[`should match the snapshot after data has been loaded 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Kristin Patrick
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Edit user
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Kristin Patrick"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Edit user"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
    data-qa-sel="page-header-edit-user"
  >
    <h1
      class="page-header__heading "
    >
      <span
        class="page-header__pre-heading"
      >
        Personal details for
      </span>
      Kristin Patrick
    </h1>
  </div>
  <div
    class="grid"
  >
    <div
      data-g="12 md:6 lg:4"
    >
      <form
        novalidate=""
      >
        <div
          class="input"
        >
          <label
            class="input__label"
            for="emailAddress"
          >
            Email address
          </label>
          <input
            autocomplete="off"
            class="input__input"
            data-qa-sel="email-input-edit-user"
            id="emailAddress"
            name="emailAddress"
            required=""
            type="email"
            value="kristin.patrick@example.com"
          />
        </div>
        <div
          class="input"
        >
          <label
            class="input__label"
            for="firstName"
          >
            First name
          </label>
          <input
            autocomplete="off"
            class="input__input"
            data-qa-sel="firstname-input-edit-user"
            id="firstName"
            maxlength="100"
            minlength="2"
            name="firstName"
            pattern="^((?![<>]).)*$"
            required=""
            type="text"
            value="Kristin"
          />
        </div>
        <div
          class="input"
        >
          <label
            class="input__label"
            for="lastName"
          >
            Last name
          </label>
          <input
            autocomplete="off"
            class="input__input"
            data-qa-sel="lastname-input-edit-user"
            id="lastName"
            maxlength="100"
            minlength="2"
            name="lastName"
            pattern="^((?![<>]).)*$"
            required=""
            type="text"
            value="Patrick"
          />
        </div>
        <div
          class="form-group-container form-group-container--no-border"
        >
          <fieldset
            class="form-group"
          >
            <legend
              class="form-group__legend"
            >
              Audience insight community membership
            </legend>
            <div
              class="radio"
            >
              <input
                checked=""
                class="radio__input"
                data-qa-sel="optIn-radio-edit-user"
                id="audienceInsight_optIn"
                name="audienceInsight"
                type="radio"
              />
              <label
                class="radio__label"
                for="audienceInsight_optIn"
              >
                Opt in
              </label>
            </div>
            <div
              class="radio"
            >
              <input
                class="radio__input"
                data-qa-sel="optOut-radio-edit-user"
                id="audienceInsight_optOut"
                name="audienceInsight"
                type="radio"
              />
              <label
                class="radio__label"
                for="audienceInsight_optOut"
              >
                Opt out
              </label>
            </div>
          </fieldset>
          <button
            class="btn btn--cta"
            data-qa-sel="save-button-edit-user"
            type="submit"
          >
            Save profile
          </button>
          <a
            class="btn btn--secondary"
            href="/users/1"
          >
            Cancel
          </a>
        </div>
      </form>
    </div>
  </div>
</div>
`;

exports[`should show error message when useFetch get call returns 401 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Edit user
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Edit user"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when useFetch get call returns 500 error 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Edit user
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Edit user"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;

exports[`should show error message when useFetch patch fails 1`] = `
<div>
  <nav
    aria-label="Breadcrumbs"
    role="navigation"
  >
    <ol
      class="breadcrumbs"
    >
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-administration-link"
          href="/"
        >
          Administration
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          data-qa-sel="breadcrumb-user-link"
          href="/users"
        >
          Users
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <a
          href="/users/1"
        >
          Error
        </a>
      </li>
      <li
        class="breadcrumbs__crumb"
      >
        <span>
          Edit user
        </span>
      </li>
    </ol>
  </nav>
  <script
    type="application/ld+json"
  >
    {
  "@context": "http://schema.org",
  "@type": "BreadcrumbList",
  "itemListElement": [
    {
      "@type": "ListItem",
      "position": 1,
      "item": {
        "@id": "/",
        "name": "Administration"
      }
    },
    {
      "@type": "ListItem",
      "position": 2,
      "item": {
        "@id": "/users",
        "name": "Users"
      }
    },
    {
      "@type": "ListItem",
      "position": 3,
      "item": {
        "@id": "/users/1",
        "name": "Error"
      }
    },
    {
      "@type": "ListItem",
      "position": 4,
      "item": {
        "name": "Edit user"
      }
    }
  ]
}
  </script>
  <div
    class="page-header"
  >
    <h1
      class="page-header__heading "
    >
      Error
    </h1>
  </div>
  <div
    class="alert alert--error"
    role="alert"
  >
    <h2>
      Whoops
    </h2>
    <p>
      Sorry, there's been an error!
    </p>
  </div>
</div>
`;
